cef/vshctrl/virtual_pbp.c

```diff
@@ -0,0 +1,1336 @@
+				memcpy(pbp, &g_caches[i], sizeof(*pbp));
+				g_referenced[i] = 1;
+
+				return 0;
+			}
+		}
+	}
+
+	return -23;
+}
+
+static int build_vpbp(VirtualPBP *vpbp) {
+	int ret, i;
+	u32 off;
+	#ifdef DEBUG
+	printk("Need to build vpbp %s\n", vpbp->name);
+	#endif
+	memset(vpbp->header, 0, sizeof(vpbp->header));
+	memset(vpbp->sects, 0, sizeof(vpbp->sects));
+	vpbp->enabled = 1;
+	vpbp->file_pointer = 0;
+	vpbp->header[0] = 0x50425000; // PBP magic
+	vpbp->header[1] = 0x10000; // version
+
+	// fill vpbp offsets
+	off = 0x28;
+
+	ret = isoOpen(vpbp->name);
+
+	if (ret < 0) {
+		#ifdef DEBUG
+		printk("%s: isoOpen -> %d\n", __func__, ret);
+		#endif
+		ret = add_cache(vpbp);
+		return ret;
+	}
+
+	for (i=0; i<NELEMS(pbp_entries); ++i) {
+		vpbp->header[i+2] = off;
+
+		if (pbp_entries[i].enabled) {
+			PBPSection *sec = &vpbp->sects[i];
+
+			ret = isoGetFileInfo(pbp_entries[i].name, &sec->size, &sec->lba);
+
+			if (ret < 0) {
+				if (i == 0) {
+					// no PARAM.SFO?
+					// then it's a bad ISO
+					isoClose();
+
+					return -36;
+				} else {
+					continue;
+				}
+			}
+
+			if (i == 0) {
+				off += sizeof(virtualsfo);
+			} else {
+				off += sec->size;
+			}
+		}
+	}
+
+	vpbp->pbp_total_size = vpbp->header[9];
+	get_iso_file_size(vpbp->name, &vpbp->iso_total_size);
+	ret = add_cache(vpbp);
+	#ifdef DEBUG
+	printk("%s: add_cache -> %d\n", __func__, ret);
+	#endif
+	isoClose();
+
+	return ret;
+}
+
+void *oe_realloc(void *ptr, int size) {
+	void *p;
+
+	p = vsh_malloc(size);
+
+	if(p != NULL && ptr != NULL) {
+		memcpy(p, ptr, size);
+		vsh_free(ptr);
+	}
+
+	return p;
+}
+
+static VirtualPBP *vpbp_realloc(VirtualPBP *vpbp, int size) {
+	VirtualPBP *v;
+
+	v = (VirtualPBP*) oe_realloc(vpbp, size * sizeof(vpbp[0]));
+
+	return v;
+}
+
+static int get_header_section(VirtualPBP *vpbp, u32 remaining, void *data) {
+	int re;
+
+	re = MIN(remaining, vpbp->header[2] - vpbp->file_pointer);
+	memcpy(data, vpbp->header+vpbp->file_pointer, re);
+
+	return re;
+}
+
+static int get_sfo_section(VirtualPBP *vpbp, u32 remaining, void *data) {
+	int re, ret;
+	void *buf, *buf_64;
+	char sfotitle[64];
+	char disc_id[12];
+	u32 parental_level = 1;
+
+	buf = vsh_malloc(SECTOR_SIZE+64);
+
+	if (buf == NULL) {
+		#ifdef DEBUG
+		printk("%s: buf cannot allocate\n", __func__);
+		#endif
+		return -13;
+	}
+
+	buf_64 = PTR_ALIGN_64(buf);
+	ret = isoRead(buf_64, vpbp->sects[0].lba, 0, SECTOR_SIZE);
+
+	if (ret < 0) {
+		#ifdef DEBUG
+		printk("%s: isoRead -> 0x%08X\n", __func__, ret);
+		#endif
+		vsh_free(buf);
+		return -37;
+	}
+
+	ret = get_sfo_string(buf_64, "TITLE", sfotitle, sizeof(sfotitle));
+
+	if (ret < 0) {
+		vsh_free(buf);
+		return ret;
+	}
+
+	ret = get_sfo_string(buf_64, "DISC_ID", disc_id, sizeof(disc_id));
+
+	if (ret < 0) {
+		vsh_free(buf);
+
+		return ret;
+	}
+
+	get_sfo_u32(buf_64, "PARENTAL_LEVEL", &parental_level);
+
+	get_sfo_u32(buf_64, "HRKGMP_VER", &(vpbp->opnssmp_type));
+
+	vsh_free(buf);
+	memcpy(virtualsfo+0x118, sfotitle, 64);
```

Comment:
ugh. this is exactly why virtualsfo were created. to not deal with magic numbers/offsets.